### PR TITLE
Bidirectional

### PR DESCRIPTION
**Pull request by LikHait**
Aquantia actively uses iperf3 for our AQC 10G products verification and testing.
Must have testing in our case is bidirect traffic runs. We suffer from iperf3 missing this feature builtin.

Here is our proposal on implementation of --bidir (bidirectional) option for iperf.

We've tried to minimize the impact and reuse existing iperf3 apis an maximum.

Short overview of the solution:

* iperf_test: `int sender` is replaced by `enum iperf_mode mode ` with 3 states:
```
enum iperf_mode {
	SENDER = 1,
	RECEIVER = 0,
	BIDIRECTIONAL = -1
};
```

* iperf_stream: add new flag "sender"

* Now streams are created depending on mode:
    * stream.sender == 1 if test->mode is SENDER
    * stream.sender == 0 if test->mode is RECEIVER
    * if test->mode is BIDIRECTIONAL we double number of streams, half has `sender == 1` and half `sender == 0`. In future it'll be possible to create asymmetric bidir flows.

* `iperf_new_stream (struct iperf_test * test, int s)` changed to
   `iperf_new_stream (struct iperf_test * test, int s, int sender)`

* Summary information output is changed for each type of iperf_stream (senders and receivers)

--------------------
### Examples

 `iperf3 -c localhost --bidir -P 4`

Interval output:
```
[  5][TX-C]   9.00-10.00  sec   584 MBytes  4.90 Gbits/sec    0   1.87 MBytes       
[  7][TX-C]   9.00-10.00  sec   584 MBytes  4.90 Gbits/sec    0   1.87 MBytes       
[  9][TX-C]   9.00-10.00  sec   584 MBytes  4.90 Gbits/sec    0   1.87 MBytes       
[ 11][TX-C]   9.00-10.00  sec   582 MBytes  4.89 Gbits/sec    0   2.12 MBytes       
[SUM][TX-C]   9.00-10.00  sec  2.28 GBytes  19.6 Gbits/sec    0             
[ 13][RX-C]   9.00-10.00  sec   453 MBytes  3.80 Gbits/sec                  
[ 15][RX-C]   9.00-10.00  sec   453 MBytes  3.80 Gbits/sec                  
[ 17][RX-C]   9.00-10.00  sec   453 MBytes  3.80 Gbits/sec                  
[ 19][RX-C]   9.00-10.00  sec   453 MBytes  3.80 Gbits/sec                  
[SUM][RX-C]   9.00-10.00  sec  1.77 GBytes  15.2 Gbits/sec     
```  

Sum output:          
```
[ ID][Role] Interval           Transfer     Bitrate         Retr
[  5][TX-C]   0.00-10.00  sec  5.83 GBytes  5.01 Gbits/sec    0             sender
[  5][TX-C]   0.00-10.02  sec  5.83 GBytes  5.00 Gbits/sec                  receiver
[  7][TX-C]   0.00-10.00  sec  5.83 GBytes  5.01 Gbits/sec    0             sender
[  7][TX-C]   0.00-10.02  sec  5.83 GBytes  5.00 Gbits/sec                  receiver
[  9][TX-C]   0.00-10.00  sec  5.83 GBytes  5.01 Gbits/sec    0             sender
[  9][TX-C]   0.00-10.02  sec  5.83 GBytes  5.00 Gbits/sec                  receiver
[ 11][TX-C]   0.00-10.00  sec  5.83 GBytes  5.01 Gbits/sec    0             sender
[ 11][TX-C]   0.00-10.02  sec  5.83 GBytes  4.99 Gbits/sec                  receiver
[SUM][TX-C]   0.00-10.00  sec  23.3 GBytes  20.0 Gbits/sec    0             sender
[SUM][TX-C]   0.00-10.02  sec  23.3 GBytes  20.0 Gbits/sec                  receiver
[ 13][RX-C]   0.00-10.00  sec  4.62 GBytes  3.96 Gbits/sec    1             sender
[ 13][RX-C]   0.00-10.02  sec  4.61 GBytes  3.95 Gbits/sec                  receiver
[ 15][RX-C]   0.00-10.00  sec  4.62 GBytes  3.96 Gbits/sec    1             sender
[ 15][RX-C]   0.00-10.02  sec  4.61 GBytes  3.95 Gbits/sec                  receiver
[ 17][RX-C]   0.00-10.00  sec  4.62 GBytes  3.96 Gbits/sec    1             sender
[ 17][RX-C]   0.00-10.02  sec  4.61 GBytes  3.95 Gbits/sec                  receiver
[ 19][RX-C]   0.00-10.00  sec  4.62 GBytes  3.96 Gbits/sec    1             sender
[ 19][RX-C]   0.00-10.02  sec  4.61 GBytes  3.95 Gbits/sec                  receiver
[SUM][RX-C]   0.00-10.00  sec  18.5 GBytes  15.9 Gbits/sec    4             sender
[SUM][RX-C]   0.00-10.02  sec  18.4 GBytes  15.8 Gbits/sec                  receiver
```

--------------------
### Comparison with legacy runs

We did a verification on our Aquantia AQC107 10G NICs.
We compared legacy usecase (using two servers for bidirect) with `--bidir`.

#### Two servers run:
iperf3 -c server -R -P 4 -t 100
```
[ ID] Interval           Transfer     Bitrate         Retr
[SUM]   0.00-100.03 sec   109 GBytes  9.38 Gbits/sec    0             sender
[SUM]   0.00-100.00 sec   109 GBytes  9.38 Gbits/sec                  receiver
```

iperf3 -c 169.254.177.210 -p 5202 -P 4 -t 100 
```
[ ID] Interval           Transfer     Bitrate         Retr
[SUM]   0.00-100.00 sec   109 GBytes  9.39 Gbits/sec    0             sender
[SUM]   0.00-100.03 sec   109 GBytes  9.39 Gbits/sec                  receiver
```

#### New, bidir run
iperf3 -c 169.254.177.210 --bidir -P 4 -t 100
```
[ ID][Role] Interval           Transfer     Bitrate         Retr
[SUM][TX-C]   0.00-100.00 sec   109 GBytes  9.37 Gbits/sec    0             sender
[SUM][TX-C]   0.00-100.03 sec   109 GBytes  9.37 Gbits/sec                  receiver
[SUM][RX-C]   0.00-100.00 sec   109 GBytes  9.39 Gbits/sec    1             sender
[SUM][RX-C]   0.00-100.03 sec   109 GBytes  9.39 Gbits/sec                  receiver
```

We've added an extra column in output, Role: `[TX-C]`. This helps to distinguish results data.
Same data is available in updated json output.

### CODE REVIEW

File: src/iperf_api.c [hunk 1]
```diff
@@ -439,11 +439,11 @@ iperf_set_test_role(struct iperf_test *ipt, char role)
...
     ipt->role = role;
     if (role == 'c')
-	ipt->sender = 1;
+	ipt->mode = SENDER;
     else if (role == 's')
-	ipt->sender = 0;
+	ipt->mode = RECEIVER;
     if (ipt->reverse)
-        ipt->sender = ! ipt->sender;
+        ipt->mode = ! ipt->mode;
```
Comment (cail) [inline src/iperf_api.c]: thats enum here, can't use `!`

File: src/iperf_api.c [hunk 2]
```diff
@@ -463,8 +463,12 @@ void
 iperf_set_test_reverse(struct iperf_test *ipt, int reverse)
 {
     ipt->reverse = reverse;
+    if (ipt->role == 'c')
+        ipt->mode = SENDER;
+    else
+        ipt->mode = RECEIVER;
     if (ipt->reverse)
-        ipt->sender = ! ipt->sender;
+        ipt->mode = ! ipt->mode;
```
Comment (cail) [inline src/iperf_api.c]: same

File: src/iperf_api.c [hunk 3]
```diff
@@ -727,6 +741,7 @@ iperf_parse_arguments(struct iperf_test *test, int argc, char **argv)
         {"length", required_argument, NULL, 'l'},
         {"parallel", required_argument, NULL, 'P'},
         {"reverse", no_argument, NULL, 'R'},
+        {"bidirectional", no_argument, NULL, OPT_BIDIRECTIONAL},
```
Comment (cail) [inline src/iperf_api.c]: `--bidir` is way better?

**Review by cail — CHANGES_REQUESTED** [review]

**Comment by LikHait** [comment]
* Implemented --bidir (bidirectional) option for iperf.
* iperf_test: int sender is replaced with enum test_mode mode with SENDER, RECEIVER and BIDIRECTIONAL states.
* iperf_stream: add new variable "sender"
* Now streams are created depending on mode:
```
	sender == 1 if test-> mode == SENDER
	sender == 0 if test-> mode == RECEIVER
	sender == 1 and sender == 0 if test-> mode == BIDIRECTIONAL
```
* `iperf_new_stream (struct iperf_test * test, int s) => iperf_new_stream (struct iperf_test * test, int s, int sender)`
* Summary information is output for each type of iperf_stream (senders and receivers)

**Review by cail — APPROVED** [review]

**Pull request merged**